packages/nextjs/hooks/scaffold-eth/useScaffoldEventHistory.ts

```diff
@@ -56,95 +56,89 @@ export const useScaffoldEventHistory = <
-          newEvents.push({
-            log: logs[i],
-            args: logs[i].args,
-            block:
-              blockData && logs[i].blockHash === null
-                ? null
-                : await publicClient.getBlock({ blockHash: logs[i].blockHash as Hash }),
-            transaction:
-              transactionData && logs[i].transactionHash !== null
-                ? await publicClient.getTransaction({ hash: logs[i].transactionHash as Hash })
-                : null,
-            receipt:
-              receiptData && logs[i].transactionHash !== null
-                ? await publicClient.getTransactionReceipt({ hash: logs[i].transactionHash as Hash })
-                : null,
+        if (blockNumber >= fromBlockUpdated) {
+          const logs = await publicClient.getLogs({
+            address: deployedContractData?.address,
+            event,
+            args: filters as any,
+            fromBlock: fromBlockUpdated,
+            toBlock: blockNumber,
           });
-        }
-        if (events && typeof fromBlock === "undefined") {
+          setFromBlockUpdated(blockNumber + 1n);
+
+          const newEvents = [];
+          for (let i = logs.length - 1; i >= 0; i--) {
+            newEvents.push({
+              log: logs[i],
+              args: logs[i].args,
+              block:
+                blockData && logs[i].blockHash === null
+                  ? null
+                  : await publicClient.getBlock({ blockHash: logs[i].blockHash as Hash }),
+              transaction:
+                transactionData && logs[i].transactionHash !== null
+                  ? await publicClient.getTransaction({ hash: logs[i].transactionHash as Hash })
+                  : null,
+              receipt:
+                receiptData && logs[i].transactionHash !== null
+                  ? await publicClient.getTransactionReceipt({ hash: logs[i].transactionHash as Hash })
+                  : null,
+            });
+          }
           setEvents([...newEvents, ...events]);
-        } else {
-          setEvents(newEvents);
+          setError(undefined);
         }
-        setError(undefined);
+      } catch (e: any) {
+        console.error(e);
+        setEvents([]);
+        setError(e);
+      } finally {
+        setIsLoading(false);
       }
-    } catch (e: any) {
-      console.error(e);
-      setEvents(undefined);
-      setError(e);
-    } finally {
-      setIsLoading(false);
-    }
-  };
-
-  useEffect(() => {
-    readEvents(fromBlock);
+    },
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [fromBlock, enabled]);
+    [
+      blockData,
+      deployedContractData,
+      enabled,
+      eventName,
+      events,
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      JSON.stringify(filters, replacer),
+      fromBlockUpdated,
+      publicClient,
+      receiptData,
+      transactionData,
+    ],
+  );
 
   useEffect(() => {
     if (!deployedContractLoading) {
```

Comment:
With the updated version, when there is an error, it keeps spamming the `readEvents` function, and we keep getting more and more errors in loop.

Should we add `&& !error` to this if condition?

Screenshot of the error spam:
![image](https://github.com/scaffold-eth/scaffold-eth-2/assets/55535804/fb4eb67e-8bbc-445c-859e-2d7f2a6c74d7)

Reply:
Thank you Pablo, good catch! Fixed [25b4295](https://github.com/scaffold-eth/scaffold-eth-2/pull/827/commits/25b42955b14260514ce7d68d0630559cbf814750). There was a problem that in `catch` block events was set to empty array even it was already empty array and `useEffect` restarted

Reply:
Working great now, TYSM!